design/ceph/object/ceph-bucket-notification-crd.md

```diff
@@ -0,0 +1,62 @@
+# Ceph bucket notifications CRD
+
+## Overview
+
+In its new release, Ceph introduced the bucket notifications feature. It allows to send messages to various endpoints when a new event occurs on a bucket (ref: <https://docs.ceph.com/docs/master/radosgw/notifications/)>
+
+Setup of those notifications are normally done by sending HTTP requests to the RGW, either to create/delete topics pointing to specific endpoints, or create/delete bucket notifications based on those topics.
+
+This functionality eases this process by avoiding to use external tools or scripts. It is replaced by creation of CR definitions that  contain all the information necessary to create topics and/or notifications, which the rook operator processes.
+
+## Goals
+
+Creates a CRD for topics and a CRD for notifications, defining all the necessary and optional information for the various endpoints.
+
+Extends the rook operator to handle the CRs that would be submitted by users.
+
+## Implementation
+
+The CR for a topic configuration takes this form:
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: Topic
+metadata:
+  name: # name of the topic
+  namespace: # namespace where topic belongs
+Spec:
+  Endpoint: #(mandatory) URI of an endpoint to send push notification to
+  Port: #(optional) port used to communicate, if empty uses the default ones
+  http:
+    verify-ssl: #indicate whether the server certificate is validated by the client or not (`true` by default)
+  amqp:
+    vhost: # defaults to `/`
+    ackLevel: # none/routable/broker, optional (default - broker)
+    amqp-exchange: # exchanges must exist and be able to route messages based on topics
+  kafka:
+    use-ssl: # secure connection will be used for connecting with the broker (`false` by default)
+    ca-location: # this specified CA will be used, instead of the default one, to authenticate the broker
+    ackLevel: # none/broker, optional (default - broker)
+```
+
+The CR for bucket notification takes this form:
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: BucketNotification
```

Comment:
If something belongs to a bucket, seems like it should be specified in an OBC instead of creating a new CRD. It would have to go under the [additionalConfig](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/object-bucket-claim-retain.yaml#L14) section which might not be ideal, but it's a consideration. And of course anything considered for the OBC should be considered in the new [k8s community design](https://github.com/kubernetes/enhancements/pull/1383).

Reply:
@travisn Yeah I agree it may need present in OBC ideally, if I go with that plan then redo all the work in upcoming `BR` part of current COSI design. But if we do it as seperate CRD then both OBC and COSI consume it very easily. Any thoughts?

P.S for COSI it can be included part of `BucketClass` not with `BucketRequest`. So in that case here it will be part of `StorageClass` than OBC. And we don't get have bucket level granularity IMO